### ISSUE TITLE
Unexpected simulation behaviour - `on_load` callback does not execute as expected

### ISSUE THREAD

**Issue by nikbpetrov** [jspsych/jsPsych#2859]
## Problem statement
When the experiment is run in simulation mode with `jsPsych.simulate()`, but a trial's `simulate` option is set to `false`, the trial's `on_load` function does **NOT** execute.

## Expected behaviour:
`on_load` to execute.

## Context
`on_load` callback executes in those cases, as expected:

- the experiment is run in non-simulation mode via `jsPsych.run()`
- the experiment is run in simulation mode but a trial's `simulate` option is set to `true`

## Reproducible example
Using, the `jspsych-html-keyboard-response.html` example, modify slightly:

```javascript
  var jsPsych = initJsPsych({
    default_ITI: 250, 
    on_finish: function() {
      jsPsych.data.displayData();
    }
  });

  var trial_1 = {
    type: jsPsychHtmlKeyboardResponse,
    stimulus: '<p style="color: red; font-size: 48px; font-weight: bold;">GREEN</p>',
    choices: ['y', 'n'],
    prompt: '<p>Does the color match the word? (y or n)</p>',
    simulation_options: {
      simulate: false
    },
    on_load: function() {
      console.log('This does NOT run when jsPsych.simulate() is executed but the trial\'s simulate option is set to false.')
    }
  };

  var trial_2 = {
    type: jsPsychHtmlKeyboardResponse,
    stimulus: '<p style="color: red; font-size: 48px; font-weight: bold;">RED</p>',
    choices: ['y', 'n'],
    trial_duration: 5000,
    prompt: '<p>Does the color match the word? (y or n; 5s time limit)</p>'
  };

  let timeline = [trial_1, trial_2]

  const SIMULATE = true
  const SIMULATE_TRIAL_DURATION = 1500
  if (SIMULATE) {
    const simulation_options = {
      default: {
        data: {
          rt: SIMULATE_TRIAL_DURATION
        }
      }
    }
    jsPsych.simulate(timeline, "visual", simulation_options)
  } else {
    jsPsych.run(timeline);
  }
```

## Possible cause
I've honed in the problem to the definition of the `load_callback` variable in `jspsych.js`
```javascript
const load_callback = () => {
    console.log('This should log in jsPsych.simulate() and simulate option of a trial set to false but it does not')
    if (typeof trial.on_load === "function") {
        trial.on_load();
    }
    // call any on_load functions for extensions
    if (Array.isArray(trial.extensions)) {
        for (const extension of trial.extensions) {
            this.extensions[extension.type.info.name].on_load(extension.params);
        }
    }
};
```

If you add a log statement inside this variable, it does not show when `jsPsych.simulate()` is run but a trial's `simulate` option is set to `false`. It appears in all other scenarios. Should mean that the callback does not get properly registered and afterwards never executes. There might be some weirness with the logic on the next lines within the `doTrial` function, but I can't tell easily.

**Comment by bjoluc** [comment]
Thanks @nikbpetrov for opening the issue. I'm tempted to leave this broken until v8 is released – guessing that it's fixed by #2858. If you have some time available, feel free to check whether `on_load` is executed in #2858; otherwise I'll look into that in a free moment.

**Comment by nikbpetrov** [comment]
@bjoluc The problem is indeed solved as described after testing with #2858, though now if a trial is run in simulation mode, the `on_load` callback is executed **twice**. This does not happen for non-simulated trials (either by not calling `jsPsych.simulate()` or by setting the trial's specific `simulate` option to `false`). To reproduce the error is the same code as above.

**Comment by bjoluc** [comment]
> though now if a trial is run in simulation mode, the on_load callback is executed twice

Oh no :see_no_evil: :bug: Thanks for checking! Looking into this now.

**Comment by bjoluc** [comment]
@nikbpetrov Not sure if I'm doing anything wrong, but I can't reproduce it :monocle_face:

I also stripped the repro down to the code below to see if  I was missing anything important:

```js
const timeline = [
  {
    type: jsPsychHtmlKeyboardResponse,
    stimulus: "foo",
    on_load: () => console.log("Will this run twice?"),
  },
];

const jsPsych = initJsPsych();

if (true) {
  jsPsych.simulate(timeline, "visual", { default: { data: { rt: 1500 } } });
} else {
  jsPsych.run(timeline);
}

```

**Comment by nikbpetrov** [comment]
@bjoluc You will have to excuse my primitive way of sharing my example, but here you go - https://drive.google.com/drive/folders/11apcv9W6kuaf-96VWWKMb3g3mI0CI-qn?usp=share_link

I forked the #2858 branch, installed all dependencies and built it as described in the [documentation](https://www.jspsych.org/7.3/developers/configuration/). I wonder whether the fact that I built it has anything to do with any differences. Still, I do see that I use the latest version of the code after building (despite the fact that the version number at the top is not yet changed). Not sure if it makes any difference but I use node 16.10.0 and npm 7.24.0

**Comment by bjoluc** [comment]
Thanks @nikbpetrov – you did everything right. I just failed to notice the little :two: next to the console message in my browser :facepalm:

My mistake was to let jsPsych invoke `on_load` when `simulate` doesn't return a promise – just like how the `trial` method works. Fixed in 0b55a324. Thank you for looking into this @nikbpetrov :heart:
